pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -0,0 +1,210 @@
+// CalculateEvenPodsSpreadPriority computes a score by checking through the topologySpreadConstraints
+// that are with WhenUnsatisfiable=ScheduleAnyway (a.k.a soft constraint).
+// The function works as below:
+// 1) In all nodes, calculate the number of pods which match <pod>'s soft topology spread constraints.
+// 2) Sum up the number to each node in <nodes> which has corresponding topologyPair present.
+// 3) Finally normalize the number to 0~10. The node with the highest score is the most preferred.
+// Note: Symmetry is not applicable. We only weigh how incomingPod matches existingPod.
+// Whether existingPod matches incomingPod doesn't contribute to the final score.
+// This is different with the Affinity API.
+func CalculateEvenPodsSpreadPriority(pod *v1.Pod, nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo, nodes []*v1.Node) (schedulerapi.HostPriorityList, error) {
+	result := make(schedulerapi.HostPriorityList, len(nodes))
+	// return if incoming pod doesn't have soft topology spread constraints.
+	constraints := getSoftTopologySpreadConstraints(pod)
+	if len(constraints) == 0 {
+		return result, nil
+	}
+
+	t := newTopologySpreadConstraintsMap(len(nodes))
+	t.initialize(pod, nodes)
+
+	allNodeNames := make([]string, 0, len(nodeNameToInfo))
+	for name := range nodeNameToInfo {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	errCh := schedutil.NewErrorChannel()
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
+		nodeInfo := nodeNameToInfo[allNodeNames[i]]
+		node := nodeInfo.Node()
+		if node == nil {
+			return
+		}
+		// (1) `node` should satisfy incoming pod's NodeSelector/NodeAffinity
+		// (2) All topologyKeys need to be present in `node`
+		if !predicates.PodMatchesNodeSelectorAndAffinityTerms(pod, node) ||
+			!predicates.NodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+			return
+		}
+		// It's enough to use topologyKey as the "key" of the map.
+		matchCount := make(map[string]int64)
+		for _, existingPod := range nodeInfo.Pods() {
+			podLabelSet := labels.Set(existingPod.Labels)
+			// Matching on constraints is calculated independently.
+			for _, constraint := range constraints {
+				match, err := predicates.PodMatchesSpreadConstraint(podLabelSet, constraint)
+				if err != nil {
+					errCh.SendErrorWithCancel(err, cancel)
+					return
+				}
+				if match {
+					matchCount[constraint.TopologyKey]++
+				}
+			}
+		}
+		// Keys in t.podCounts have been ensured to contain "filtered" nodes only.
+		for _, constraint := range constraints {
+			tpKey := constraint.TopologyKey
+			pair := topologyPair{key: tpKey, value: node.Labels[tpKey]}
+			// For each <pair>, all matched nodes get the credit of summed matchCount.
+			// And we add matchCount to <t.total> to reverse the final score later.
+			for _, nodeName := range t.topologyPairToNodeNames[pair] {
+				atomic.AddInt64(t.podCounts[nodeName], matchCount[tpKey])
+				atomic.AddInt64(&t.total, matchCount[tpKey])
```

Comment:
At the end of the day, total is a constant for all nodes (we normalize between min and max score). So it's meaningless when calculating the priority.

Reply:
Below is how I think about this problem:

If we don't have a `total` value, we only know the number of matching pods on each node, that's not enough. As the number of matching pods sort of represents the credit it "owe", but without `total`, we don't have a global picture how many credits the whole system "owe"s, hence can't know the "owed" credit ratio for each node.

For example, two nodes are matched with 5/6 pods, some other priorities use an algorithm like this: find the max(6), min defaults to 0, then the first node gets a score 2 (= 10-50/6), and the second gets 0!! (= 10-60/6) That's not a good normalization.

With `total`, we can give better math: reverse the matching to be 6/5 (11 -5or6), then give the final score: the first node gets 10 (= 60/6), and the second gets 8 (= 50/6).

Reply:
Right, if you normalize between best and worst node, then it doesn't serve any purpose. But you are (now?) normalizing between 0 and best node.

Reply:
Yes, the old commit had `minCount`, but that also ends up with 0. So given that matching number can't be negative, I assumed `minCount` to be 0.